### PR TITLE
LLM: add CPU More-Data-Types and Save-Load examples

### PR DESCRIPTION
**Pull request by JinBridger**
## Description

Add CPU `More-Data-Types` and `Save-Load` examples.

### CODE REVIEW

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+## Example: Predict Tokens using `generate()` API
+In the example [generate.py](./generate.py), we show a basic use case of low-bit optimizations on a Llama2 model to predict the next N tokens using `generate()` API, with BigDL-LLM low-bit optimizations.
+### 1. Install
+We suggest using conda to manage environment:
+```bash
+conda create -n llm python=3.9
+conda activate llm
+
+pip install bigdl-llm[all] # install bigdl-llm with 'all' option
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md:16]: pip install --pre --upgrade bigdl-llm[all]

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# BigDL-LLM More Data Types for Low-Bit Inference on Large Language Model
+
+In this example, we show how to apply BigDL-LLM low-bit optimizations (including INT8/INT5/INT4) to Llama2 model, and then run inference on the optimized low-bit model.
+
+## 0. Requirements
+To run this example with BigDL-LLM, we have some recommended requirements for your machine, please refer to [here](../../README.md#system-support) for more information.
+
+## Example: Predict Tokens using `generate()` API
+In the example [generate.py](./generate.py), we show a basic use case of low-bit optimizations on a Llama2 model to predict the next N tokens using `generate()` API, with BigDL-LLM low-bit optimizations.
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md:9]: I think we should provide more descriptions about the usage of applying other low-bit optimization at the beginning.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+```bash
+conda create -n llm python=3.9
+conda activate llm
+
+pip install bigdl-llm[all] # install bigdl-llm with 'all' option
+```
+
+### 2. Run
+```
+python ./generate.py
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md:21]: 'python ./generate.py --low-bit sym_int8' may be better.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+### RESPONSE:
+"""
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Predict Tokens using `generate()` API for Llama2 model')
+    parser.add_argument('--repo-id-or-model-path', type=str, default="meta-llama/Llama-2-7b-chat-hf",
+                        help='The huggingface repo id for the Llama2 (e.g. `meta-llama/Llama-2-7b-chat-hf` and `meta-llama/Llama-2-13b-chat-hf`) to be downloaded'
+                             ', or the path to the huggingface checkpoint folder')
+    parser.add_argument('--low-bit', type=str, default="sym_int4",
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py:37]: should we use "sym_int4" by default ? We provide lots of examples of int4 optimization in Model folder.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    args = parser.parse_args()
+    model_path = args.repo_id_or_model_path
+    low_bit = args.low_bit
+
+    # Load model
+    model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True)
+    
+    # With only one line to enable BigDL-LLM optimization on model
+    model = optimize_model(model, low_bit=low_bit)
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py:53]: I think we could add some descriptions about `low_bit` param here.

File: python/llm/example/CPU/PyTorch-Models/Save-Load/README.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# BigDL-LLM Save/Load Low-Bit Model
+
+In this example, we show how to save/load model with BigDL-LLM low-bit optimizations (including INT8/INT5/INT4), and then run inference on the optimized low-bit model.
+
+## 0. Requirements
+To run this example with BigDL-LLM, we have some recommended requirements for your machine, please refer to [here](../../README.md#system-support) for more information.
+
+## Example: Predict Tokens using `generate()` API
+In the example [generate.py](./generate.py), we show a basic use case of low-bit optimizations on a Llama2 model to predict the next N tokens using `generate()` API, with BigDL-LLM low-bit optimizations.
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/Save-Load/README.md:9]: I think we should update descriptions about this example. And we could mention that saving and loading operations are platform-independent.

File: python/llm/example/CPU/PyTorch-Models/Save-Load/README.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# BigDL-LLM Save/Load Low-Bit Model
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/Save-Load/README.md:1]: maybe "Save/Load Low-Bit Models with BigDL-LLM Optimizations" ?

File: python/llm/example/CPU/PyTorch-Models/Save-Load/generate.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import torch
+import time
+import argparse
+from bigdl.llm import optimize_model
+from bigdl.llm.optimize import low_memory_init, load_low_bit
+from transformers import AutoModelForCausalLM, LlamaTokenizer, TextGenerationPipeline
```
Comment (plusbang) [inline python/llm/example/CPU/PyTorch-Models/Save-Load/generate.py:22]: I think we don't need `TextGenerationPipeline` here.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
+# BigDL-LLM More Data Types for Low-Bit Inference on Large Language Model
```
[the remark this answers is not in the document]
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md]: Done.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md (hunk 1 above)
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md]: Done.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md (hunk 2 above)
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/More-Data-Types/README.md]: Done.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py (hunk 4 above)
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py]: Changed to "sym_in8" instead.

File: python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py (hunk 5 above)
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/More-Data-Types/generate.py:55]: Added.

File: python/llm/example/CPU/PyTorch-Models/Save-Load/README.md (hunk 6 above)
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/Save-Load/README.md:9]: Already done as you said.

File: python/llm/example/CPU/PyTorch-Models/Save-Load/README.md (hunk 7 above)
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/Save-Load/README.md]: Done.

File: python/llm/example/CPU/PyTorch-Models/Save-Load/README.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+We suggest using conda to manage environment:
+```bash
+conda create -n llm python=3.9
+conda activate llm
+
+pip install bigdl-llm[all] # install bigdl-llm with 'all' option
+```
+
+### 2. Run
+If you want to save low-bit optimized model, run:
```
[the remark this answers is not in the document]
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/Save-Load/README.md]: Done.

File: python/llm/example/CPU/PyTorch-Models/Save-Load/generate.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+# you could tune the prompt based on your own model,
+# here the prompt tuning refers to https://huggingface.co/georgesung/llama2_7b_chat_uncensored#prompt-style
+LLAMA2_PROMPT_FORMAT = """### HUMAN:
+{prompt}
+
+### RESPONSE:
+"""
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Transformer save_load example')
```
[the remark this answers is not in the document]
Comment (JinBridger) [reply python/llm/example/CPU/PyTorch-Models/Save-Load/generate.py]: Done.

**Comment by JinBridger** [comment]
Already adjusted according to comments.

**Review by plusbang — APPROVED** [review]

**Pull request merged**